database/src/main/java/com/firebase/ui/database/FirebaseIndexArray.java

```diff
@@ -41,8 +42,11 @@ public FirebaseIndexArray(Query keyRef, Query dataRef) {
     @Override
     public void cleanup() {
         super.cleanup();
-        for (Map.Entry<Query, ValueEventListener> entry : mRefs.entrySet()) {
+        final Iterator<Map.Entry<Query, ValueEventListener>> it = mRefs.entrySet().iterator();
+        while (it.hasNext()) {
+            final Map.Entry<Query, ValueEventListener> entry = it.next();
             entry.getKey().removeEventListener(entry.getValue());
+            it.remove();
```

Comment:
Yes, that case is specifically mentioned in the javadoc for Map.entrySet().  https://developer.android.com/reference/java/util/Map.html#entrySet()